### DISCUSSION TITLE
Trying to write a loader that compiles a file and returns its result as string

### DISCUSSION THREAD

**Discussion by LukasBombach** [webpack/webpack#16816] [Q&A] [answered]
I am trying to write a loader with which I can compile a requested modules and return its compiled source as string. I am doing this to find a way to inline some imported JavaScript in a React / Next.js app with SSR:

```ts
// file_a.js

export const a = "A";
```

```ts
// file_b.js

import { a } from "./file_a";

console.log(a);
```

```tsx
import sourceAsString from "my-compiling-loader!./file_b";

/**
 * sourceAsString should now be the compiled source of file_a and file_b as a string
 * <script dangerouslySetInnerHTML={{ __html: 'const a = "A";console.log(a);' }} />;
 */
export InlineScript = () => <script dangerouslySetInnerHTML={{ __html: sourceAsString }} />;
```

I tried doing this with a custom webpack loader and a combination of the

* `raw-loader` and the
* `importModule` api

the idea is to

* request a resource as string with the `raw-loader`
* instead of just returning the resource, I am returning the compiled file

So I was expecting to get the compiled source (file_b and file_a as single compiled source code) but all I get is the source code of the file I requested with the `raw-loader` in  the first place (so here the source code of `file_b`)

```js
// my-compiling-loader.js

module.exports = async function (content, map, meta) {
  if (/GET_CONTENTS/.test(this.request)) {
    return content;
  }

  if (/DO_COMPILE/.test(this.request)) {
    const result = await this.importModule(this.request + "?GET_CONTENTS");
    return result.default || result;
  }

  return "module.exports = require('raw-loader!" + this.request + "?DO_COMPILE');";
};
```

**note**

I am doing a bit of a hack here, instead of having 3 loaders to resolve my resource in the steps `raw-loader > importModule > content` I doing sort of a dirty hack with `?DO_COMPILE` and `?GET_CONTENTS` to reuse the same loader

⬇️ 

now I am wondering how I can request the contents of a fully compiled file

**Comment by alexander-akait** [comment, marked as answer]
> now I am wondering how I can request the contents of a fully compiled file

Can you clarify? Compiled after what?

**Reply by LukasBombach** [reply]
So, in my first example, when I would request 

```
import sourceAsString from "my-compiling-loader!./file_b";
```

`sourceAsString` would equal the source code of the file_b and its bundled (that's what I mean with compiled) dependencies. So, for illustration, something like

```
sourceAsString === "const a = 'A';\nconsole.log(a);"
```

**Reply by alexander-akait** [reply]
So you want to replace import on static code?

**Reply by LukasBombach** [reply]
No I need to ask you to clarify :)

I want to import code as a static string, yes. But not the original source, but the transformed and bundled source of a given file.

**Reply by LukasBombach** [reply]
I have done something similar already

https://github.com/LukasBombach/next-inline-script

we have that in production at https://www.t-online.de/ but the version that we use here (the one from this repo) has a few problems that come from using a child compiler. So I am looking for a way that does not use a child compiler)

**Reply by alexander-akait** [reply]
> I want to import code as a static string, yes. But not the original source, but the transformed and bundled source of a given file.

In this case you still need extra compilation - you can use multi compier mode (bundle your script firstly, don't forget to set `dependecies`, so compilation of an inline script will be first) and when use assets module load this script https://webpack.js.org/guides/asset-modules/#replacing-inline-loader-syntax

**Reply by LukasBombach** [reply]
By multicompiler mode you mean an array of configurations for webpack?

Like this?

```
webpack([...])
```

**Reply by alexander-akait** [reply]
Yes :+1:

**Reply by LukasBombach** [reply]
I see, thank you.

There is one last approach I am thinking about: You know, in the end I am trying to inline some "compiled" JavaScript in a `<script />` tag via React and SSR.

And I am doing this as part of a Next.js application. 

So I wondering if this would work: A loader could add the requested file as an entry point its compiler. And at build time, instead of returning the file's contents, it would return code that would read the source code of the compiled code at runtime.

So something like this:

```js
// my-loader.js (mock code)

module.exports = function () {

  this._compiler.addEntry(this.resourcePath);

  return `module.exports = require("fs").readFileSync(/*output path for this file's entry*/);`;
}
```

Leaving aside questions like server vs. client side access to the file system (I have some ideas how to deal with that inside next js), it seem feasable to me that

* at build time, by dynamically adding an entry point, I would get the source code assets that I need
* at build time, I would generate code that will instruct the runtime to read that asset
* and at run time, during ssr, the generated entry point would be read from the file system and be used in my application
  (again I am assuming here that I have a workaround for dealing with the client, in short, next has a multicompiler, one for ssr one for the browser and I would generated different code for each one)


My issue with that approach is, that I cannot figure out how (or if) it is possible to dynamically add entries to the compiler from a loader

**Reply by alexander-akait** [reply]
> My issue with that approach is, that I cannot figure out how (or if) it is possible to dynamically add entries to the compiler from a loader

Yeah, it was not desig for this, but you can try use this https://github.com/webpack/webpack-dev-server/blob/master/lib/Server.js#L676, don't forget to add this to dependecies, i.e. `this.addDependency(pathToFile);` for cache invalidation

**Reply by alexander-akait** [reply]
i.e. you can get `webpack` using `this._compiler.webpack`, compiler you already have

**Reply by LukasBombach** [reply]
So, I came up with that. Do you think that this seems right?

(I tried it along with some other versions, but next js does seem to add some strange behavior on top of webpack which makes it hard for me to validate if this code is correct)

```js
const path = require("node:path");

/**
 * @this {import('webpack').LoaderContext<{}>}
 * @type {import('webpack').LoaderDefinitionFunction}
 */
function compileToStringLoader(content, sourceMap, additionalData) {
  const callback = this.async();

  const webpack = this._compiler.webpack;
  const compiler = this._compiler;
  const context = compiler.context;
  const entry = path.relative(context, this.resourcePath);

  new webpack.EntryPlugin(context, entry, {
    name: "my-entry-test-name",
    filename: "my-entry-test-filename.js",
  }).apply(compiler);

  this.addDependency(this.resourcePath);

  callback(null, content, sourceMap, additionalData);
}

module.exports = compileToStringLoader;
```

**Reply by alexander-akait** [reply]
To be honestly, code is valid should work, but I don't test it, do you try to run it?

**Reply by LukasBombach** [reply]
I tried it, it _does_ work—but apparently only in `dev` mode. Next.js seems to be doing something else to interfere.

I will do a separate deep dive for this and maybe go another route.

Well but, thank you for your help. I really do appreciate what you are doing here, I see that you are really supporting a lot of issues  as a maintainer here (some others of mine as well) and I know that this is all voluntary and want you to know that this is really helpful and I am thankful that you take the time for this.

**Reply by alexander-akait** [reply]
Thank you, anyway if you will create small reproducible test repo with an example with your loader and describe a logic in readme (like actual/expected), I can look and try to help you, yeah, nextjs has own plugin and some stuff, so not all things which are working with pure webpack config will work with nextjs, but mostly should work

**Reply by LukasBombach** [reply]
That's ok thank you. This is for my day job and if I go down this road any further, it seems at that I'd have to rely on next-internals, which I can't justify from a professional / business point of view.

I think at this point I am going to invest some more time in trying to accomplish this with a child compiler and mark this one as answered.

Again, thank you so much!
